use logger created from logger factory instead of console logger

vhatsura opened the pull request:
related to #228.

What was done:
1. Use logger created from `ILoggerFactory` inside `Elastic.Apm` infrastructure.
2. Additional dependency to `Microsoft.Extensions.Logging.Abstractions`

Open questions:
1. `IApmLogger` has `Level` property in interface definition. In case of asp.net core philosophy, configuration per logger category is splitted from `ILogger` functionality.
2. `Elastic.Apm.AspNetCore` project doesn't have extensions for `IServiceCollection`. Instead of directly creation of some instances, there are can be configured through dependency injection. It's allow to simplify configuration in the future and replace some parts of `Elastic.Apm` by users.

@gregkalapos, what do you think about my changes? Is it aligned with your thoughts?

vhatsura commented on src/Elastic.Apm.AspNetCore/ApmMiddlewareExtension.cs:
```diff
@@ -35,7 +36,9 @@ public static class ApmMiddlewareExtension
-			var logger = ConsoleLogger.Instance;
+			var loggerFactory = builder.ApplicationServices.GetService(typeof(ILoggerFactory)) as ILoggerFactory;
+
+			var logger = new AspNetCoreLogger(loggerFactory);
```
fixed

vhatsura commented on src/Elastic.Apm.AspNetCore/ApmMiddlewareExtension.cs:
```diff
@@ -35,7 +36,9 @@ public static class ApmMiddlewareExtension
 		)
 		{
-			var logger = ConsoleLogger.Instance;
+			var loggerFactory = builder.ApplicationServices.GetService(typeof(ILoggerFactory)) as ILoggerFactory;
```
fixed

vhatsura commented on src/Elastic.Apm.AspNetCore/AspNetCoreLogger.cs:
```diff
@@ -0,0 +1,39 @@
+
+		private readonly ILogger _logger;
+
+		public LogLevel Level { get; } = LogLevel.Error;
```
Agreed with you. Error was a simple solution for draft pull request. Also, I agreed with the 2nd option (use log level from standard asp.net core config section). However, fetch logger configuration isn't a simple task in asp.net core due to it can be configured in various way. By this reson, I iterate over `IsEnabled` method to find out min level which is enabled for logger.

gregkalapos replied:
> However, fetch logger configuration isn't a simple task 

Yes, I also tried to come up with something, but that is kind of not part of the API. One thing comes to my mind which may be a problem: Is the loglevel changeable on the fly while the application is running? Atm the code caches the value in the `AspNetCoreLogger` construction, and I wonder if this can change. 
On the other hand reading it always, when the `LogLevel` property is accessed sounds dangerous in terms of overhead (if we go there, we can just measure this).

But overall I think we are good to processed here, this should not block you, we may rethink this part, but it's ok for now.

vhatsura commented:
@gregkalapos, thank you for review, all coments are relevant and was fixed. Could you look again, please? If all is ok, I will start to cover changes by unit tests

gregkalapos commented:
Thank you @vhatsura, 
I think it's ok to proceed here, feel free to continue and also feel free to ping me if you have other questions. 

>  Also, I will appreciate if you specify which docs need to updated with configuration info for asp.net core apm agent

[This file](https://github.com/elastic/apm-agent-dotnet/blob/master/docs/setup.asciidoc) is I think a good place to describe this. Feel free to simply add it to the end.

vhatsura commented:
Hi, @gregkalapos. I almost done with unit tests (tests AspNetCoreLogger are left), however, have a simple question: what do you think about using [`Moq`](https://github.com/moq/moq4) library?

gregkalapos commented on src/Elastic.Apm.AspNetCore/AspNetCoreLogger.cs:
```diff
@@ -0,0 +1,39 @@
+
+		public AspNetCoreLogger(ILoggerFactory loggerFactory)
+		{
+			_logger = loggerFactory?.CreateLogger("Apm") ?? throw new ArgumentNullException(nameof(loggerFactory));
```
Could we maybe make it `ElasticApm` instead of `Apm`? That's be the same as the category we use in `appsettings.json`, and also we'd not take up the word that is used industry wide for ourself.

vhatsura replied:
changed

gregkalapos commented on docs/configuration.asciidoc:
```diff
@@ -36,20 +36,22 @@ With this you can use any configuration source that you configured on the `IConf
 
 NOTE: By simply calling `app.UseElasticApm()` without the overload, the agent will read configurations only from environment variables.
 
+NOTE: For better user experience, `LogLevel` for internal apm logger is used from standard `Logging` section inside `appsettings.json` with `Apm` category name.
```
These changes only apply to ASP.NET Core (since we activate the new logger in the `ApmMiddlewareExtension`), so I'd try to make it clear in the docs.

I propose to add this hint to the beginning: `For better user experience in case of ASP.NET Core, `LogLevel` for internal apm …`

vhatsura replied:
added

gregkalapos commented on docs/configuration.asciidoc:
```diff
@@ -36,20 +36,22 @@ With this you can use any configuration source that you configured on the `IConf
 
+NOTE: For better user experience, `LogLevel` for internal apm logger is used from standard `Logging` section inside `appsettings.json` with `Apm` category name.
+
 This is a typical `appsettings.json` file that contains some sample configuration. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
```
Similarly, here I'd make clear that this is an `appsettings.json` sample for ASP.NET Core.

vhatsura replied:
added

gregkalapos commented on docs/configuration.asciidoc:
```diff
@@ -36,20 +36,22 @@ With this you can use any configuration source that you configured on the `IConf
   "AllowedHosts": "*",
   "ElasticApm":
     {
-      "LogLevel":  "Debug",
```
Additionally I'd add another block below this one to explain how to use `ElasticApm:LogLevel` when ASP.NET Core is not involved. 

Something like this:

In case you don't activate the agent with the `UseElasticApm()` method (typically non-ASP.NET Core scenario) then you can set the loglevel of the agent with the `ElasticApm:LogLevel` setting, as shown in the following `appsettings.json` file:

[and then use the original file]

(_feel free to adapt or improve the text if you feel something is not ok_)

vhatsura replied:
added

vhatsura commented:
Hey, @gregkalapos !
Thanks for review. I updated `configuration` document in documentation and changed category name for logger. Also, I wrote unit tests for `AspNetCoreLogger` with `Moq`. It is what I needed from Moq. If you decide that it would be better to have such unit tests with own mocks, then I will update such tests, but IMHO, with `Moq` such tests is more elegant

bmorelli25 commented on docs/configuration.asciidoc:
```diff
@@ -36,7 +36,29 @@ With this you can use any configuration source that you configured on the `IConf
 NOTE: By simply calling `app.UseElasticApm()` without the overload, the agent will read configurations only from environment variables.
 
-This is a typical `appsettings.json` file that contains some sample configuration. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
+NOTE: For better user experience in case of ASP.NET Core, `LogLevel` for internal apm logger is used from standard `Logging` section inside `appsettings.json` with `Apm` category name.
```
I don't love the idea of having two _NOTEs_ in a row:
<img width="784" alt="Screen Shot 2019-07-01 at 2 17 11 PM" src="https://user-images.githubusercontent.com/5618806/60467248-f2493200-9c0a-11e9-9b91-21ee18cd43b6.png">

Does this new information need to be in a note? Also, it could be lack of .NET knowledge, but I'm not exactly sure what this is trying to say? @gregkalapos do you?

gregkalapos replied:
Yes, I do. This changes how to set the loglevel for the agent. In case of ASP.NET Core, here is how to do it:
```
{
	  "Logging": {
	    "LogLevel": {
	      "Default": "Warning",
	      "ElasticApm": "Debug"
	    }
	  },
	  "AllowedHosts": "*",
	  "ElasticApm":
	    {
	      "ServerUrls":  "http://myapmserver:8200",
	      "TransactionSampleRate": 1.0
	    }
	}
```

The `Logging` part is standard and also used by the ASP.NET Core logger - so we (the APM Agent) log to the same logger.

Now, if you don't use ASP.NET Core (e.g. you have a command line application) we can't register our logger with the default logger (since there isn't really one), so in that case here is how you set the log level (it's moved to the `ElasticApm` section:

```
{
	  "Logging": {
	    "LogLevel": {
	      "Default": "Warning",
	    }
	  },
	  "AllowedHosts": "*",
	  "ElasticApm":
	    {
	      "ServerUrls":  "http://myapmserver:8200",
	      "TransactionSampleRate": 1.0,
              "LogLevel" : "Debug"
	    }
	}
```

Probably there is a better way to phrase this; happy to take suggestions :)

gregkalapos replied:
Btw. now that I see the text again, we should change 
> `appsettings.json` with `Apm` category name 

 to 

> `appsettings.json` with `ElasticApm` category name` 

We changed that after one of my comments.

bmorelli25 replied:
This definitely helps, thanks Greg! I'm wondering if we should put this under a "sample config" section... here's what I'm thinking:

```asciidoc
[float]
[[sample-config]]
==== Sample configuration file

Below is a sample `appsettings.json` configuration file for a typical ASP.NET Core application. There are two important takeaways:
1. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
2. With ASP.NET Core, we recommend setting `LogLevel for the internal APM logger in the standard `Logging` section with the `ElasticApm` category name.

[source,js]
----
{
  "Logging": {
    "LogLevel": {
      "Default": "Warning",
      "ElasticApm": "Debug"
    }
  },
  "AllowedHosts": "*",
  "ElasticApm":
    {
      "ServerUrls":  "http://myapmserver:8200",
      "TransactionSampleRate": 1.0
    }
}
----

In certain scenarios -- typically when you're not using ASP.NET Core -- you may choose not to active the agent with the `UseElasticApm()` method.
In this case, you can set the log level of the agent with `ElasticApm:LogLevel`, as shown in the following `appsettings.json` file:

[source,js]
----
{
  "Logging": {
    "LogLevel": {
      "Default": "Warning"
    }
  },
  "AllowedHosts": "*",
  "ElasticApm":
    {
      "LogLevel":  "Debug",
      "ServerUrls":  "http://myapmserver:8200",
      "TransactionSampleRate": 1.0
    }
}
----
```

gregkalapos replied:
Yeah, I think the structure is very good! 

Some comments:

> With ASP.NET Core, we recommend setting `LogLevel` for the internal APM logger in the standard `Logging`

It's not a recommendation, it's the only way to set it, if you do it the other way (like the 2. example above, it'll be ignored). 

> In certain scenarios -- typically when you're not using ASP.NET Core -- you may choose not to active the agent with the `UseElasticApm()` method.

This is very good, I'd just change a small thing: the part with "you may choose not to active the agent with..."  should be a different. It's not really a choice, if you are not on ASP.NET Core you can't really call the `UseElasticApm` method, so it should be something like "you typically don't activate the agent with..."

Overall I think with this structure the doc is good and easy to understand. On the other hand I'm still not sure this is a good move, it's still hard to explain this, it can be confusing. I think it'd be better to just use the same in both cases. Unfortunately I think there is no way to fetch the loglevel from `"ElasticApm":  { "LogLevel":  "..."}` when we use `ILoggerFactory`.

So I see 2+1 options:

- We go with the improved docs 
- We continue thinking about this
- (and +1: we go with the improved docs and reiterate on this and see if there is a better way to always configure the loglevel the same way)

I'd choose the last one. 

@vhatsura let us know what you think.

bmorelli25 replied:
Thanks again Greg! I've added your recommendations and updated the docs below. If you and @vhatsura revisit this and are able to fetch the loglevel in the same place, let me know and I can update the docs accordingly.

```asciidoc
[float]
[[sample-config]]
==== Sample configuration file

Below is a sample `appsettings.json` configuration file for a typical ASP.NET Core application. There are two important takeaways:
1. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
2. With ASP.NET Core, you must set `LogLevel for the internal APM logger in the standard `Logging` section with the `ElasticApm` category name.

[source,js]
----
{
  "Logging": {
    "LogLevel": {
      "Default": "Warning",
      "ElasticApm": "Debug"
    }
  },
  "AllowedHosts": "*",
  "ElasticApm":
    {
      "ServerUrls":  "http://myapmserver:8200",
      "TransactionSampleRate": 1.0
    }
}
----

In certain scenarios -- like when you're not using ASP.NET Core -- you wont activate the agent with the `UseElasticApm()` method.
In this case, you can set the log level of the agent with `ElasticApm:LogLevel`, as shown in the following `appsettings.json` file:

[source,js]
----
{
  "Logging": {
    "LogLevel": {
      "Default": "Warning"
    }
  },
  "AllowedHosts": "*",
  "ElasticApm":
    {
      "LogLevel":  "Debug",
      "ServerUrls":  "http://myapmserver:8200",
      "TransactionSampleRate": 1.0
    }
}
----
```

gregkalapos replied:
Thanks @bmorelli25. I think this is a fairly important decision - I thought about it and opened https://github.com/elastic/apm-agent-dotnet/issues/338 with a suggestion - we'll decide there. Will ping you once we came up with something.

vhatsura replied:
updated

bmorelli25 commented on docs/configuration.asciidoc:
```diff
@@ -36,7 +36,29 @@ With this you can use any configuration source that you configured on the `IConf
-This is a typical `appsettings.json` file that contains some sample configuration. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
+NOTE: For better user experience in case of ASP.NET Core, `LogLevel` for internal apm logger is used from standard `Logging` section inside `appsettings.json` with `Apm` category name.
+
+This is a typical `appsettings.json` file for ASP.NET Core application that contains some sample configuration. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
```
```suggestion
Below is a sample `appsettings.json` configuration file for a typical ASP.NET Core application. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
```

bmorelli25 commented on docs/configuration.asciidoc:
```diff
@@ -36,7 +36,29 @@ With this you can use any configuration source that you configured on the `IConf
+}
+----
+
+In case you don't activate the agent with the `UseElasticApm()`` method (typically non-ASP.NET Core scenario) then you can set the log level of the agent with the `ElasticApm:LogLevel` setting, as shown in the following `appsettings.json` file:
```
```suggestion
In certain scenarios -- typically when you're not using ASP.NET Core -- you may choose not to active the agent with the `UseElasticApm()` method.
In this case, you can set the log level of the agent with `ElasticApm:LogLevel`, as shown in the following `appsettings.json` file:
```

gregkalapos commented on sample/SampleAspNetCoreApp/appsettings.json:
```diff
@@ -1,12 +1,12 @@
     "LogLevel": {
-      "Default": "Warning"
+      "Default": "Warning",
+      "Apm": "Error"
```
Should be `ElasticApm`

vhatsura replied:
fixed

kuisathaverat commented:
jenkins run the test please

SergeyKleyman commented on src/Elastic.Apm.AspNetCore/Extensions/LoggerExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+		{
+			if (logger == null) throw new ArgumentNullException(nameof(logger));
+
+			if (logger.IsEnabled(Microsoft.Extensions.Logging.LogLevel.Trace)) return LogLevel.Trace;
```
I think we should replace `Level` property in `IApmLogger` by `IsEnabled` with the same semantics as `Microsoft.Extensions.Logging`'s `ILogger.IsEnabled`. This is the only way `IApmLogger.Level` is currently used anyway and it will avoid adding code like this. @gregkalapos WDYT?

gregkalapos replied:
I don't really see the end result of it - we log the `LogLevel`, so we should know that (it's part of `void Log<TState>(LogLevel level, TState state, Exception e, Func<TState, Exception, string> formatter)`). But maybe it's just I don't see the whole thing here.

SergeyKleyman replied:
I think there is some misunderstanding: log level that is included in the log record is derived from the level of the logging statement that emitted the record. For example, log records emitted by the following statement `logger.Debug()?.Log(...)` will have level `Debug`. OTOH `Level` property in `IApmLogger` controls which records are allowed to be emitted based on their level so it functions exactly like Microsoft.Extensions.Logging's `ILogger.IsEnabled`. Other logging libraries do it the same way as Microsoft.Extensions.Logging: for example [Serilog's `ILogger.IsEnabled`](https://github.com/serilog/serilog/blob/dev/src/Serilog/ILogger.cs#L190) and [NLog's `ILogger.IsTraceEnabled`](https://github.com/NLog/NLog/blob/master/src/NLog/ILogger.cs#L49)
Because of this discrepancy between the way Microsoft.Extensions.Logging does it and the way we currently specified `IApmLogger` interface (by forcing `Level` property), the bridge between `IApmLogger` and Microsoft.Extensions.Logging's `ILogger`, implemented in this PR, has to "guess" the minimal level that is not filtered out by repeatedly calling `ILogger.IsEnabled` as this code does. This is both wasteful from performance point of view and very hackish code because `IApmLogger.Level` is used only by [`LoggingExtensions.IfLevel`](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/Logging/IApmLoggingExtensions.cs#L32) and it used exactly like `ILogger.IsEnabled`
```
logger.Level <= level ? new MaybeLogger(logger, level) : (MaybeLogger?)null;
```
It's like parsing a number from a string just to print it. But in this case it's even worse because it also prevents supporting dynamically changing the level. If we replace `IApmLogger`'s `Level` by `IsEnabled` then `AspNetCoreLogger.IsEnabled` from this PR can just call `_logger.IsEnabled` and automatically support more use cases (dynamically changing the level in particular) with less code - no need to guess level anymore and thus `LoggerExtensions.GetMinLogLevel` is not necessary.

gregkalapos replied:
Ok, I see... so you want an `bool IsEnabled(LogLevel)` method (Or `bool IsDebugEnabled()`) methods - not an `bool IsEnabled` property.

So we'd have this:

```
public interface IApmLogger
{
	bool IsEnabled(LogLevel level);		
}
```

and this:
```
internal class AspNetCoreLogger : IApmLogger
{
   private readonly ILogger _logger;

   public bool IsEnabled(LogLevel level)
	=> _logger.IsEnabled(Convert(level));
//...more coee
}
```

Right?

SergeyKleyman replied:
Yes, exactly. Although I'm not sure what you mean by `bool IsEnabled` property - do you mean NLog's approach with `bool IsDebugEnabled` etc. I'm not sure why they did that way (performance ?) - I would just go with Microsoft.Extensions.Logging and Serilog's `ILogger.IsEnabled` approach. It's hidden behind `LoggingExtensions.Trace()`, `LoggingExtensions.Debug()` etc. so even if in the future we change our minds and decide to go with NLog's `IsDebugEnabled` approach we will just need to change the way `LoggingExtensions.Trace()` is implemented but the rest of the agent won't be affected.

gregkalapos replied:
> Although I'm not sure what you mean by bool IsEnabled property 

No.. that was my original interpretation of you original message, like I thought you want a property called `IsEnabled` that returns a bool - which was obviously not what you meant :) I think we understand each other now - C# code snippet helps ;)

vhatsura replied:
It would be nice to have consistent API with `Microsoft.Extensions.Logging`, however, I think it's not a goal of this PR. For me, it would be better to have another PR with replacing `Level` property to `IsEnabled`. @gregkalapos, @SergeyKleyman, what do you think about it?

SergeyKleyman replied:
@vhatsura I would recommend to make this change in this PR but it's not a blocker for me so the final decision is yours.

The reason I recommend to make this change in this PR is because it's a classic case for refactoring - you want to implement a new feature but the current design is too rigid to accommodate it. Of course you can hack around the current design but I prefer the approach of refactoring the existing code and then implementing the new feature on top thus avoiding hacks. 

But in this case the refactoring required is very localized so postponing it doesn't matter that much that is why it's just a recommendation from me and the final decision is yours.

gregkalapos replied:
I'd prefer to do it in another PR, mainly because it's a breaking change. I think having a breaking change as part of a feature PR isn't optimal - And since we can avoid it, it take that route.

vhatsura replied:
Got it. Will do changes in another PR

vhatsura commented on src/Elastic.Apm.AspNetCore/ApmMiddlewareExtension.cs:
```diff
@@ -62,6 +64,13 @@ public static class ApmMiddlewareExtension
 		}
 
+		internal static IApmLogger GetApmLogger(this IServiceProvider serviceProvider)
+		{
```
fixed

vhatsura commented on test/Elastic.Apm.AspNetCore.Tests/AspNetCoreLoggerTests.cs:
```diff
@@ -0,0 +1,50 @@
+	{
+		[Fact]
+		public void AspNetCoreLoggerShouldThrowExceptionWhenLoggerFactoryIsNull()
+		{
```
fixed

vhatsura commented on sample/SampleAspNetCoreApp/appsettings.json:
```diff
@@ -1,12 +1,12 @@
   },
   "AllowedHosts": "*",
   "ElasticApm": {
-    "LogLevel": "Error",
```
added

vhatsura commented on src/Elastic.Apm.AspNetCore/AspNetCoreLogger.cs:
```diff
@@ -0,0 +1,39 @@
+
+		public AspNetCoreLogger(ILoggerFactory loggerFactory)
+		{
+			_logger = loggerFactory?.CreateLogger("ElasticApm") ?? throw new ArgumentNullException(nameof(loggerFactory));
```
changed

vhatsura commented:
@gregkalapos, @SergeyKleyman, @bmorelli25, it seems I fixed all your comments. Can you look into changes once again?

gregkalapos commented:
> @gregkalapos, @SergeyKleyman, @bmorelli25, it seems I fixed all your comments. Can you look into changes once again?

Yes, will go over it soon one last time. I think we'll be good to merge 👍  

Thanks for all the work @vhatsura, very much appreciated!

bmorelli25 commented on docs/configuration.asciidoc:
```diff
@@ -36,7 +36,34 @@ With this you can use any configuration source that you configured on the `IConf
+==== Sample configuration file
+
+Below is a sample `appsettings.json` configuration file for a typical ASP.NET Core application. There are two important takeaways:
+1. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
```
```suggestion

1. The part below `ElasticApm` is fetched by the agent if the corresponding `IConfiguration` is passed to the agent.
```
